packages/Code/src/Code.tsx

```diff
@@ -0,0 +1,218 @@
+      splitString.shift();
+    }
+
+    // If the last line is blank, remove the last line of code.
+    // This is a similar issue to the one above.
+    while (stringFragmentIsInvalid(splitString[splitString.length - 1])) {
+      splitString.pop();
+    }
+
+    return {
+      content: splitString.join('\n'),
+      lineCount: splitString.length,
+    };
+  }, [snippet]);
+}
+
+const whiteSpace = 12;
+
+const wrapperStyle = css`
+  overflow-x: auto;
+  border: 1px solid;
+  border-left-width: 3px;
+  border-radius: 4px;
+  padding: ${whiteSpace}px;
+  margin: 0;
+  position: relative;
+`;
+
+const wrapperStyleWithLineNumbers = css`
+  padding-left: ${whiteSpace * 3.5}px;
+`;
+
+const wrapperStyleWithWindowChrome = css`
+  padding-top: ${windowChromeHeight + whiteSpace}px;
+  border-left-width: 1px;
+`;
+
+function getWrapperVariantStyle(variant: Variant): string {
+  const colors = variantColors[variant];
+
+  return css`
+    border-color: ${colors['01']};
+    background-color: ${colors['00']};
+    color: ${colors['03']};
+  `;
+}
+
+interface Props extends SyntaxProps {
+  /**
+   * Shows line numbers in preformatted code blocks.
+   *
+   * default: `false`
+   */
+  showLineNumbers?: boolean;
+
+  /**
+   * Shows window chrome for code block;
+   *
+   * default: `false`
+   */
+  showWindowChrome?: boolean;
+
+  /**
+   * Renders a file name or other descriptor for a block of code
+   */
+  chromeTitle?: string;
+
+  /**
+   * When true, whitespace and line breaks will be preserved.
+   *
+   * default: `true`
+   * */
+  multiline?: boolean;
+}
+
+type DetailedElementProps<T> = React.DetailedHTMLProps<
+  React.HTMLAttributes<T>,
+  T
+>;
+
+/**
+ * # Code
+ * 
+ * React Component that outputs single-line and multi-line code blocks.
+ * 
+ * ```
+<Code>Hello world!</Code>
+	```
+ * ---
+ * @param props.children The string to be formatted.
+ * @param props.className An additional CSS class added to the root element of Code.
+ * @param props.multiline When true, whitespace and line breaks will be preserved. Default: `true`
+ * @param props.lang The language used for syntax highlighing. Default: `auto`
+ * @param props.variant Determines if the code block is rendered with a dark or light background. Default: 'light'
+ * @param props.showLineNumbers When true, shows line numbers in preformatted code blocks. Default: `false`
+ */
+function Code({
+  children = '',
+  className,
+  multiline = true,
+  lang = Lang.Auto,
+  variant = Variant.Light,
+  showLineNumbers = false,
+  showWindowChrome = false,
+  chromeTitle = '',
+  ...rest
+}: Props) {
+  const wrapperClassName = cx(
+    wrapperStyle,
+    getWrapperVariantStyle(variant),
+    {
+      [wrapperStyleWithLineNumbers]: multiline && showLineNumbers,
+      [wrapperStyleWithWindowChrome]: showWindowChrome,
+    },
+    className,
+  );
+
+  const { content, lineCount } = useProcessedCodeSnippet(children);
+
+  if (!multiline) {
```

Comment:
Could probably unify the resulting JSX rendered by moving conditionals around the Root element type and elsewhere. Seems not too bad to DRY up.

Reply:
The reason I didn't do this is that I wasn't sure how to deal with the typings for `...rest` in that case. Right now I'm doing 
```js
{...(rest as DetailedElementProps<HTMLDivElement>)}
```
and 
```js
{...(rest as DetailedElementProps<HTMLPreElement>)}
```
 in each code path respectively, and I'm not entirely sure how to do this otherwise....

Ideas?

Reply:
Mmmm, no idea how to do that either. 

Could probably do the cast in its own if statement or ternary along with the element as Root.

At this point I'm not sure it's super helpful though, to do this refactor.

If anything it makes more sense to have the contents be in a variable so there's no risk of the contents becoming out of sync. That's my primary concern, that the contents of the component can become out of sync.

Reply:
Contents are in a variable – it's destructured from `useProcessedCodeSnippet()`.

Reply:
I meant the contents inside the div and pre tags.

Reply:
Alright, I can render the window chrome and Syntax within a shared variable.